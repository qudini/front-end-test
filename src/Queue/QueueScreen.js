import React, {useEffect, useCallback, useState} from 'react';
import { connect, useDispatch } from 'react-redux';
import { fetchQueueData } from "../mockApi";
import md5 from 'md5';
import { setCustomers } from "./actions/customerActions";
import Name from "./components/Name";
import Time from "./components/Time";
import Content from "./components/Content";
import CustomerCard from "./components/CustomerCard";
import ProfilePicture from "./components/ProfilePicture";
import { SET_CUSTOMERS } from "./actions/actionTypes";
import Filter from "./components/Filter";
import Heading from "./components/Heading";

const defaultPerson = '/images/defaultPerson.jpeg';

const QueueScreen = ({ customers }) => {
    const [filterValue, setFilterValue] = useState('');
    const dispatch = useDispatch();

    const fetchGravatarImages = useCallback(async (customers) => {
        if (!customers || !Array.isArray(customers)) {
            return;
        }

        // Create an array to hold the updated customers
        const updatedCustomers = [];

        // Use Promise.all to wait for all image fetches to complete
        await Promise.all(customers.map(async (customer) => {
            const { emailAddress } = customer.customer;

            // Check if emailAddress is not null and not undefined
            if (emailAddress && typeof emailAddress === 'string' && emailAddress.trim() !== '') {
                const gravatarUrl = `https://www.gravatar.com/avatar/${md5(emailAddress)}?d=identicon`;
                try {
                    const response = await fetch(gravatarUrl);
                    if (!response.ok) {
                        console.error('Error fetching Gravatar image. HTTP status:', response.status);
                        // Assign the default image URL to customers without images
                        updatedCustomers.push({ ...customer, gravatarUrl: defaultPerson });
                    } else {
                        const blob = await response.blob();
                        const imageUrl = URL.createObjectURL(blob);
                        // Add the updated customer to the array
                        updatedCustomers.push({ ...customer, gravatarUrl: imageUrl });
                    }
                } catch (error) {
                    console.error("Error fetching Gravatar image:", error);
                }
            } else {
                // Assign the default image URL to customers without valid email addresses
                updatedCustomers.push({ ...customer, gravatarUrl: defaultPerson });
            }
        }));

        // Dispatch the action with the updated customers
        dispatch({
            type: SET_CUSTOMERS,
            payload: updatedCustomers,
        });
    }, [dispatch]);

    useEffect(() => {
        const fetchData = async () => {
            try {
                const response = await fetchQueueData();
                if (!response) {
                    console.error('Error fetching data: Response is undefined');
                    return;
                }
                const json = await response.json();
                if (!json) {
                    console.error('Error fetching data: JSON is undefined');
                    return;
                }
                const fetchedCustomers = json.queueData.queue.customersToday;
                dispatch(setCustomers(fetchedCustomers));
                await fetchGravatarImages(fetchedCustomers);

            } catch (error) {
                console.error("Error fetching data:", error);
            }
        };
        fetchData()
    }, [dispatch, fetchGravatarImages]);

    const handleFilterChange = (value) => {
        setFilterValue(value);
    };

    const filteredCustomers = customers && Array.isArray(customers)
        ? customers.filter((customer) =>
            customer.customer.name.toLowerCase().includes(filterValue.toLowerCase())
        )
        : [];

    return (
        <Content>
            <Heading level={1}>Customer queue</Heading>
            <Filter filterValue={filterValue} onFilterChange={handleFilterChange} />
            {filteredCustomers.map((customer) =>
            (
                <CustomerCard key={customer.customer.id}>
                    <Name>{customer.customer.name}</Name>
                    <ProfilePicture imageUrl={customer.gravatarUrl ? customer.gravatarUrl : defaultPerson } alt={customer.customer.name}/>
                    <Time dateTime={customer.expectedTime} />
                </CustomerCard>
            ))}
        </Content>
    );
};

const mapStateToProps = (state) => ({
    customers: state.queue.customers,
});

export default connect(mapStateToProps, { setCustomers })(QueueScreen);
