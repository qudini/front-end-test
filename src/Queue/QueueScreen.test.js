// QueueScreen.test.js
import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer'
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { fetchQueueData } from '../mockApi';
import QueueScreen from './QueueScreen';

// Mocking the fetchQueueData function
jest.mock('../mockApi', () => ({
    fetchQueueData: jest.fn(),
}));

URL.createObjectURL = jest.fn();

describe('QueueScreen', () => {
    const initialState = {
        queue: {
            customers: [
                {
                    customer: {
                        id: 1,
                        name: 'John Doe',
                        emailAddress: 'john.doe@example.com',
                    },
                    gravatarUrl: 'mocked-url1',
                    expectedTime: '2023-11-20T12:00:00.000Z',
                },
                {
                    customer: {
                        id: 2,
                        name: 'Jane Doe',
                        emailAddress: 'jane.doe@example.com',
                    },
                    gravatarUrl: 'mocked-url2',
                    expectedTime: '2023-11-20T12:00:00.000Z',
                },

            ],
        },
    };
    const mockStore = configureStore();
    let store;

    beforeEach(() => {
        store = mockStore(initialState);
        global.fetch = jest.fn();

        // Mocking a successful response
        fetch.mockResolvedValueOnce({
            ok: true,
            blob: () => Promise.resolve(new Blob()),
        });
    });

    it('renders the component with customer cards', async () => {
        fetchQueueData.mockResolvedValueOnce({
            status: 200,
            json: () => Promise.resolve({ queueData: { queue: { customersToday: [] } }, status: 'ok' }),
        });

        render(
            <Provider store={store}>
                <QueueScreen />
            </Provider>
        );
        await waitFor(() => {
            expect(screen.getByText('John Doe')).toBeInTheDocument();
        });
    });

    it('handles empty customers array', async () => {
        const initialStateWithEmptyArray = {
            queue: {
                customers: [],
            },
        };
        const storeWithEmptyArray = mockStore(initialStateWithEmptyArray);

        render(
            <Provider store={storeWithEmptyArray}>
                <QueueScreen />
            </Provider>
        );

        await waitFor(() => {
            expect(screen.queryByText('John Doe')).not.toBeInTheDocument();
        });
    });
    it('handles when there is no customers array', async () => {
        const initialStateWithEmptyArray = {
            queue: {
                customers: null,
            },
        };
        const storeWithEmptyArray = mockStore(initialStateWithEmptyArray);

        render(
            <Provider store={storeWithEmptyArray}>
                <QueueScreen />
            </Provider>
        );

        await waitFor(() => {
            expect(screen.queryByText('John Doe')).not.toBeInTheDocument();
        });
    });
    it('handles error when fetching Gravatar image', async () => {
        const mockCustomer = {
            customer: {
                id: 1,
                name: 'John Doe',
                emailAddress: 'john.doe@example.com',
            },
        };

        const initialState = {
            queue: {
                customers: [mockCustomer],
            },
        };
        const store = mockStore(initialState);

        // Mock a failed response
        fetch.mockResolvedValueOnce({
            ok: false,
            status: 404,
            statusText: 'Not Found',
        });

        render(
            <Provider store={store}>
                <QueueScreen />
            </Provider>
        );

        await waitFor(() => {
            // Verify that the default image URL is used for the customer without images
            expect(screen.getByText('John Doe')).toBeInTheDocument();
            expect(screen.getByAltText('John Doe')).toHaveAttribute('src', '/images/defaultPerson.jpeg');
        });
    });


    it('filters customers by name', async () => {
        fetchQueueData.mockResolvedValueOnce({
            status: 200,
            json: () => Promise.resolve({ queueData: { queue: { customersToday: initialState.queue.customers } }, status: 'ok' }),
        });

        render(
            <Provider store={store}>
                <QueueScreen />
            </Provider>
        );

        await waitFor(() => {
            expect(screen.getByText('John Doe')).toBeInTheDocument();
        });

        await userEvent.type(screen.getByRole('textbox', {name: /filter/i}), 'John');

        await waitFor(() => {
            expect(screen.getByText('John Doe')).toBeInTheDocument();
            expect(screen.queryByText('Jane Doe')).not.toBeInTheDocument();
        });

        await userEvent.clear(screen.getByRole('textbox', {name: /filter/i}));

        await waitFor(() => {
            expect(screen.getByText('John Doe')).toBeInTheDocument();
            expect(screen.getByText('Jane Doe')).toBeInTheDocument();
        });
    });
    it('matches snapshot', () => {
        const tree = renderer.create(
            <Provider store={store}>
                <QueueScreen />
            </Provider>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
