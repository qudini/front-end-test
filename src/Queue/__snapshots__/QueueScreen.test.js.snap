// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueueScreen matches snapshot 1`] = `
.c6 {
  font-size: 20px;
  color: darkslategrey;
}

.c8 {
  font-size: 16px;
  color: darkslategrey;
}

.c0 {
  padding-left: 1em;
  padding-right: 1em;
  font-family: 'Ubuntu',sans-serif;
  display: flex;
  flex-direction: column;
  gap: 20px;
}

.c5 {
  border-style: solid;
  border-width: 1px;
  border-color: lightgray;
  border-radius: 0.2em;
  width: 300px;
  height: 370px;
  margin-bottom: 20px;
  padding-left: 20px;
  padding-right: 20px;
  padding-bottom: 10px;
}

.c7 {
  height: 290px;
  width: 290px;
  background-color: grey;
  border-top-left-radius: 0.2em;
  border-top-right-radius: 0.2em;
}

.c2 {
  padding-bottom: 10px;
  width: 342px;
}

.c3 {
  display: block;
  margin-bottom: 10px;
  font-weight: bold;
  color: darkslategrey;
}

.c4 {
  width: 100%;
  padding: 8px;
  box-sizing: border-box;
}

.c1 {
  font-weight: bold;
  margin-bottom: 10px;
  color: darkslategrey;
}

.c1 h1 {
  font-size: 2em;
}

.c1 h2 {
  font-size: 1.5em;
}

.c1 h3 {
  font-size: 1.2em;
}

<div
  className="c0"
>
  <h1
    className="c1"
  >
    Customer queue
  </h1>
  <div
    className="c2"
  >
    <label
      className="c3"
      htmlFor="filterInput"
    >
      Filter
    </label>
    <input
      className="c4"
      id="filterInput"
      onChange={[Function]}
      placeholder="Filter by name"
      type="text"
      value=""
    />
  </div>
  <div
    className="c5"
  >
    <h1
      className="c6"
    >
      John Doe
    </h1>
    <img
      alt="John Doe"
      className="c7"
      src="mocked-url1"
    />
    <h1
      className="c8"
    >
      November 20, 2023, 12:00:00 PM GMT
    </h1>
  </div>
  <div
    className="c5"
  >
    <h1
      className="c6"
    >
      Jane Doe
    </h1>
    <img
      alt="Jane Doe"
      className="c7"
      src="mocked-url2"
    />
    <h1
      className="c8"
    >
      November 20, 2023, 12:00:00 PM GMT
    </h1>
  </div>
</div>
`;
